opm/simulators/timestepping/TimeStepControl.cpp

```diff
@@ -292,83 +293,140 @@ namespace Opm
-            return std::numeric_limits<double>::max();
-        }
-        // Use an I controller after report time steps or chopped time steps
-        else if (substepTimer.currentStepNum() < 3 || substepTimer.lastStepFailed() || counterSinceFailure_ > 0)
-        {
-            if (substepTimer.lastStepFailed() || counterSinceFailure_ > 0)
-                counterSinceFailure_++;
-            if (counterSinceFailure_ > 1)
-                counterSinceFailure_ = 0;
-            const double newDt = dt * std::pow(safetyFactor_ * tolerance_ / errors_[2], 0.35);
-            if( verbose_ )
+            controllerVersion_ = 0; // "I-controller"
+            const double newDt = dt * timeStepFactor(errors_, timeSteps_);
+            if (verbose_)
+            {
                 OpmLog::info(fmt::format("Computed step size (pow): {} days", unit::convert::to( newDt, unit::day )));
+            }
             return newDt;
         }
         // Use the general third order controller for all other time steps
         else
         {
-            const std::array<double, 3> beta = { 0.125, 0.25, 0.125 };
-            const std::array<double, 2> alpha = { 0.375, 0.125 };
-            const double newDt = dt * std::pow(safetyFactor_ * tolerance_ / errors_[2], beta[0]) *
-                                      std::pow(safetyFactor_ * tolerance_ / errors_[1], beta[1]) *
-                                      std::pow(safetyFactor_ * tolerance_ / errors_[0], beta[2]) *
-                                      std::pow(timeSteps_[2] / timeSteps_[1], -alpha[0]) *
-                                      std::pow(timeSteps_[1] / timeSteps_[0], -alpha[1]);
-            if( verbose_ )
+            controllerVersion_ = 1; // "3rdOrder";
+            const double newDt = dt * timeStepFactor(errors_, timeSteps_);
+            if (verbose_)
+            {
                 OpmLog::info(fmt::format("Computed step size (pow): {} days", unit::convert::to( newDt, unit::day )));
+            }
             return newDt;
         }
     }
 
+    double General3rdOrderController::
+    timeStepFactor(const std::array<double, 3> errors, const std::array<double, 3> timeSteps) const
+    {
+        // "3rdOrder"
+        if (controllerVersion_ == 1)
+        {
+            return std::pow(safetyFactor_ * tolerance_ / errors[2], beta_[0]) *
+                   std::pow(safetyFactor_ * tolerance_ / errors[1], beta_[1]) *
+                   std::pow(safetyFactor_ * tolerance_ / errors[0], beta_[2]) *
+                   std::pow(timeSteps[2] / timeSteps[1], -alpha_[0]) *
+                   std::pow(timeSteps[1] / timeSteps[0], -alpha_[1]);
+        }
+        // "I-controller"
+        return std::pow(safetyFactor_ * tolerance_ / errors[2], 0.35);
+    }
+
     bool General3rdOrderController::
-    timeStepAccepted(const double error) const
+    timeStepAccepted(const double error, const double timeStep) const
```

Comment:
The change to making the storing and shifting of errors here is a significant change to the API, which makes it required to call timeStepAccepted() every step. Please document in header.